### DISCUSSION TITLE
Bugs fix and future features

### DISCUSSION THREAD

**Discussion by darkyne** [INGInious/problems-math#8] [Ideas]
# List of known bugs to fix and potential future features:

# BUGS:

###  subscripted variables crashing: FIXED 
This is now handled by the parser.

###  The 'i' complex constant does not act consistently: FIXED
Sometimes similar expressions using the i constant are not correctly evaluated leading to wrong results.
It may also be useful to specify its use in the documentation (for example, using I or i ?)
Now i (not I) always acts as the imaginary constant. But this means no variable can take name i (same as e)

###  Tolerance and conversions: FIXED
Sometimes problems using a tolerance margin act wrongly, especially when using constants such as 'pi' or comparisons between sqrt and numerical value.
The is_equal method was reworked and now works flowlessly.

###  Integers/floats conversions: FIXED
Some expressions when using float or integers representations are not giving equal results.
Example1:
Solution: sqrt(0.5)
Answer1: sqrt(0.5): Accepted
Answer2: sqrt(1/2): Refused
This was fixed when reworking the is_equal method.

### Clarifying the case of intervals: FIXED
Verify that answers with the form of intervals are correctly handled.
Example: -infinity,5] U [8,11] U [13,+infinity
Intervals were added with math_interval problem type.

###  similar polynomial expressions not being equal: FIXED
When encoding polynomial equations, some mathematically similar equations do not give the same result.
Example:
Solution: y(1-a)=0
Answer1: y(1-a)=0: Accepted
Answer2: y-ay=0: Refused
Answer3: 2y(1/2-a/2)=0: Refused
SOLUTION: This is due to x(a) being considered as " x of a" and not x*a.
To avoid this confusion while not removing the function feature, some user interface instructions would be a good start.
Some user instructions and conventions are now displayed with a button.

###  Clarifying the use of log and ln: FIXED
Some students may be confused with the two symbols. 
log represents log base 10 and ln represents natural log. It is also possible to define log like: log_10, log_5, log_e etc to represent log with specific base.
This is displayed via the information button.

### Clarifying the use of dot and the coma: FIXED
Should the two symbols act the same way? 
Specifying what exactly each represents may be useful in the documentation.
Right now, the '.' represents a decimal part and the ',' may be used for delimitation for example when we implement vectors.
This is displayed via the information button.

###  Variables can not be named 'e': FIXED
'e' always represents the 2.71 constant and can never be used to represent a variable.
Proposition: Give more information about the constant behaviours in documentation, for example telling that 'e' always represents the 'e' constant.
This is displayed via the information button.

###  Clarifying the case of inequality: OK
Verify that anwers with the form of inequality equation are correctly handled.
Inequalities such as x<=5 are well handled.
Inequalitie such as 1<=x<=5 are not handled with this format but are handled with interval format such as [1,5] 

###  Error message saying Error in MCQ while it is in a math question: 
When a math answer is refused because it is wrongly formatted, the error message says there is an error in MCQ.
This does not happen when the answer is wrong with correct format.

# FEATURES:

### Vector encoding: ADDED
It could be useful to allow vector encoding with a format smilar to [a, b, c]

### Matrix encoding: ADDED
It could be useful to allow matrix encoding with a format smilar to [a, b; c, d]

### Geometric point encoding: ADDED (matrix with size 1x2)
It could be useful to allow geometric point encoding with a format smilar to (x,y)
Current fix: Adding two solutions
New better fix: (a,b) is a matrix of size 1x2, it is perfect to handle a point. You can add multiple points with "adding more answers"

### Allow two modes: Expression mode or numerical mode:  NOT REQUIRED ANYMORE
Some subproblem may expect a numerical value when some other problems may expect an expression.
Using a checkbox for the two modes when encoding the answer may help to correctly evaluate the answer with the correct context. The numerical mode may cast the answer to a float to allow better evaluation in the case of tolerance margin for example.
This was added with the math_numerical problem type.
Now the is_equal method allow to correctly evaluate answers, a separation between numerical and expression answers is not required anymore.

### Partial derivates:
Partial derivates are not understood by the parser.

### Adding sets:
Add the sets from sympy, allow unions and intersections.
Example: {1, 4, 6}

### Allow solutions to be visible after a certain condition is met:
It may be useful for non-certifying tests to allow the answer to be displayed after a certain number of fails or after a certain date.
Current solution: The feedback allows to put some restructured text which can include some 'hidden until' parts.
Better solution: Allow this feature within the teacher graphical interface when configuring the subproblem so he does not have to write restructed text by his own.

### Custom symbols pannel:
It may be useful if we know the answer will include a symbol such as pi to already include it in the graphical symbols pannel available to students.

### Make the use of graphs easier:
Right now, it's relatively hard for a teacher to properly display graphs.
It may be helpful to have options for graphs in user-interface. Or maybe a specific problem type for graphs questions to avoid making the user interface heavier.

### Multiple correct answers possible:
It could be useful for some exercises to accept any answer from a set of correct solutions.
For example if we look for any root from a function not all of them.
Current fix: Specify in the context what solution to enter

### MCQ no correct answer:
If a MCQ is designed so that no answer is correct, the student still has to select one answer or it will not be submitted.
Current fix: Add a "None of the answers" choice.

### Single subproblem submission:
Some tasks may include multiple subproblems such as, for example, a mathematical problem followed by a MCQ talking about the math problem answer. If the student wants to submit only the first subproblem, he needs to answer to the following MCQ or it will not be submitted. It would be better to allow only first subproblem submission.
Current fix: The student answers randomly to the following questions just to submit the first subproblem. 

### Step-by-step exercises:
Allow for a specific task to be split into multiple subtasks representing the different steps of the initial problem.
Current solution: A link in the context, allowing the student to have access to another task with more subproblems representing the different steps.

**Comment by anthonygego** [comment]
Partial derivates are not understood by the parser neither.

**Reply by darkyne** [reply]
Added in future features :)
